File: src-tauri/src/types/collections.rs
//! World-Class Collections System - Rust Implementation
//!
//! This module provides enterprise-grade collection and workflow types that mirror
//! the TypeScript interface but with Rust's safety and performance advantages.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use uuid::Uuid;

// =============================================================================
// Core Collection Types
// =============================================================================

/// Enhanced Collection with advanced workflow capabilities
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Collection {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub tags: Vec<String>,

    // Workflow definition
    pub workflow: Vec<WorkflowStep>,

    // Global variables and configuration
    pub variables: HashMap<String, CollectionVariable>,
    pub environment: CollectionEnvironment,

    // Metadata
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub created_by: Option<String>,
    pub version: String,

    // Execution history
    pub last_run: Option<DateTime<Utc>>,
    pub run_count: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CollectionVariable {
    pub name: String,
    pub description: Option<String>,
    #[serde(rename = "type")]
    pub var_type: VariableType,
    pub default_value: Option<serde_json::Value>,
    pub required: bool,
    pub current_value: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum VariableType {
    String,
    Number,
    Boolean,
    Json,
    Secret,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CollectionEnvironment {
    pub name: String,
    pub description: Option<String>,
    pub servers: HashMap<String, Uuid>, // alias -> server_id mapping
    pub variables: HashMap<String, serde_json::Value>,
}

// =============================================================================
// Workflow System
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowStep {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,

    // Step configuration
    pub enabled: bool,
    pub continue_on_error: bool,
    pub timeout_ms: Option<u32>,

    // Dependencies
    pub depends_on: Vec<Uuid>, // Other step IDs that must complete first

    // The actual operation
    pub operation: McpOperation,

    // Variable management
    pub variable_extracts: Vec<VariableExtract>,
    pub assertions: Vec<Assertion>,

    // Execution metadata
    pub execution_order: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum McpOperation {
    Tool {
        server_alias: String,
        tool_name: String,
        parameters: HashMap<String, serde_json::Value>,
    },
    Resource {
        server_alias: String,
        resource_uri: String,
    },
    Prompt {
        server_alias: String,
        prompt_name: String,
        parameters: HashMap<String, serde_json::Value>,
    },
    Sampling {
        server_alias: String,
        messages: Vec<SamplingMessage>,
        max_tokens: Option<u32>,
        temperature: Option<f32>,
        auto_approve: Option<bool>,
    },
    Elicitation {
        server_alias: String,
        request_id: String,
        response_data: serde_json::Value,
    },
    Delay {
        duration_ms: u32,
    },
    Conditional {
        condition: String, // JavaScript expression using variables
        then_steps: Vec<Uuid>, // Step IDs to execute if true
        else_steps: Option<Vec<Uuid>>, // Step IDs to execute if false
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SamplingMessage {
    pub role: MessageRole,
    pub content: serde_json::Value, // Can be string or content array
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MessageRole {
    User,
    Assistant,
    System,
}

// =============================================================================
// Variable System
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VariableExtract {
    // Where to extract the value from
    pub source: ExtractSource,

    // JSONPath or simple property access
    pub path: String, // e.g., "$.result.token" or "response.user.id"

    // Variable to store the extracted value
    pub variable_name: String,

    // Optional transformations
    pub transform: Option<VariableTransform>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ExtractSource {
    Response,
    Status,
    Timing,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VariableTransform {
    #[serde(rename = "type")]
    pub transform_type: TransformType,
    pub options: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TransformType {
    String,
    Number,
    Boolean,
    Json,
    Base64Encode,
    Base64Decode,
    Hash,
}

// =============================================================================
// Assertion System
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Assertion {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,

    // What to assert
    #[serde(rename = "type")]
    pub assertion_type: AssertionType,

    // The assertion logic
    pub condition: AssertionCondition,

    // Behavior on failure
    pub severity: AssertionSeverity,
    pub continue_on_failure: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AssertionType {
    ResponseStatus,
    ResponseContains,
    ResponseEquals,
    ResponseJsonPath,
    Timing,
    VariableValue,
    CustomScript,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssertionCondition {
    pub operator: AssertionOperator,
    pub expected_value: serde_json::Value,
    pub actual_path: Option<String>, // JSONPath for complex response validation
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AssertionOperator {
    Equals,
    NotEquals,
    Contains,
    NotContains,
    GreaterThan,
    LessThan,
    RegexMatch,
    JsonSchema,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AssertionSeverity {
    Error,
    Warning,
    Info,
}

// =============================================================================
// Execution System
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowExecution {
    pub id: Uuid,
    pub collection_id: Uuid,
    pub collection_version: String,

    // Execution metadata
    pub started_at: DateTime<Utc>,
    pub completed_at: Option<DateTime<Utc>>,
    pub status: ExecutionStatus,

    // Results
    pub step_results: HashMap<Uuid, StepResult>,
    pub final_variables: HashMap<String, serde_json::Value>,
    pub summary: ExecutionSummary,

    // Runtime configuration
    pub environment_name: String,
    pub user_variables: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ExecutionStatus {
    Running,
    Completed,
    Failed,
    Cancelled,
    Paused,
}

/// Event emitted during workflow execution for real-time monitoring
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowExecutionEvent {
    pub execution_id: String,
    pub event_type: String,
    pub timestamp: String,
    pub step_id: Option<String>,
    pub message: Option<String>,
    pub data: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StepResult {
    pub step_id: Uuid,
    pub status: StepStatus,

    // Timing
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub duration_ms: Option<u32>,

    // Results
    pub operation_result: Option<serde_json::Value>,
    pub extracted_variables: HashMap<String, serde_json::Value>,
    pub assertion_results: Vec<AssertionResult>,

    // Error handling
    pub error: Option<String>,
    pub retry_count: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StepStatus {
    Pending,
    Running,
    Completed,
    Failed,
    Skipped,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssertionResult {
    pub assertion_id: Uuid,
    pub passed: bool,
    pub message: String,
    pub expected: Option<serde_json::Value>,
    pub actual: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExecutionSummary {
    pub total_steps: u32,
    pub completed_steps: u32,
    pub failed_steps: u32,
    pub skipped_steps: u32,

    pub total_duration_ms: u32,
    pub total_assertions: u32,
    pub passed_assertions: u32,
    pub failed_assertions: u32,

    pub servers_used: Vec<String>,
    pub operations_performed: HashMap<String, u32>, // operation_type -> count
}

// =============================================================================
// Import/Export System
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CollectionExport {
    pub format_version: String,
    pub exported_at: DateTime<Utc>,
    pub exported_by: Option<String>,

    pub collection: Collection,
    pub environments: Vec<CollectionEnvironment>,

    // Optional: Include recent execution history
    pub recent_executions: Option<Vec<WorkflowExecution>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CollectionImportOptions {
    pub merge_environments: bool,
    pub override_existing: bool,
    pub preserve_ids: bool,

    // Server mapping for when server IDs don't match
    pub server_mapping: Option<HashMap<Uuid, Uuid>>,
}

// =============================================================================
// Collection Templates System
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CollectionTemplate {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub category: String,
    pub tags: Vec<String>,

    // Template-specific fields
    pub author: Option<String>,
    pub documentation_url: Option<String>,
    pub required_server_types: Vec<String>, // e.g., ["auth-server", "api-server"]

    // The template collection (with placeholder values)
    pub template_collection: Collection,

    // Instructions for users
    pub setup_instructions: String,
    pub usage_examples: Vec<String>,
}

// =============================================================================
// Constants
// =============================================================================

pub const COLLECTION_FORMAT_VERSION: &str = "1.0.0";

impl Default for CollectionEnvironment {
    fn default() -> Self {
        Self {
            name: "Default".to_string(),
            description: Some("Default environment".to_string()),
            servers: HashMap::new(),
            variables: HashMap::new(),
        }
    }
}

impl Collection {
    /// Create a new empty collection
    pub fn new(name: String, description: Option<String>) -> Self {
        let now = Utc::now();
        Self {
            id: Uuid::new_v4(),
            name,
            description,
            tags: Vec::new(),
            workflow: Vec::new(),
            variables: HashMap::new(),
            environment: CollectionEnvironment::default(),
            created_at: now,
            updated_at: now,
            created_by: None,
            version: COLLECTION_FORMAT_VERSION.to_string(),
            last_run: None,
            run_count: 0,
        }
    }

    /// Add a workflow step to the collection
    pub fn add_step(&mut self, step: WorkflowStep) {
        self.workflow.push(step);
        self.updated_at = Utc::now();
    }

    /// Update a workflow step
    pub fn update_step(&mut self, step_id: Uuid, step: WorkflowStep) -> Result<(), String> {
        if let Some(existing_step) = self.workflow.iter_mut().find(|s| s.id == step_id) {
            *existing_step = step;
            self.updated_at = Utc::now();
            Ok(())
        } else {
            Err(format!("Step with ID {} not found", step_id))
        }
    }

    /// Remove a workflow step
    pub fn remove_step(&mut self, step_id: Uuid) -> Result<(), String> {
        let initial_len = self.workflow.len();
        self.workflow.retain(|s| s.id != step_id);

        if self.workflow.len() < initial_len {
            self.updated_at = Utc::now();
            Ok(())
        } else {
            Err(format!("Step with ID {} not found", step_id))
        }
    }

    /// Get steps in execution order, respecting dependencies
    pub fn get_execution_order(&self) -> Result<Vec<&WorkflowStep>, String> {
        let mut executed = std::collections::HashSet::new();
        let mut execution_order = Vec::new();
        let mut remaining_steps: std::collections::HashMap<Uuid, &WorkflowStep> =
            self.workflow.iter().map(|s| (s.id, s)).collect();

        // Simple dependency resolution (topological sort)
        while !remaining_steps.is_empty() {
            let mut progress_made = false;

            // Find steps with all dependencies satisfied
            let ready_steps: Vec<Uuid> = remaining_steps
                .iter()
                .filter(|(_, step)| {
                    step.depends_on.iter().all(|dep_id| executed.contains(dep_id))
                })
                .map(|(id, _)| *id)
                .collect();

            if ready_steps.is_empty() {
                return Err("Circular dependency detected in workflow steps".to_string());
            }

            // Sort ready steps by execution_order
            let mut ready_with_order: Vec<_> = ready_steps
                .into_iter()
                .map(|id| (remaining_steps[&id].execution_order, id))
                .collect();
            ready_with_order.sort_by_key(|(order, _)| *order);

            // Add ready steps to execution order
            for (_, step_id) in ready_with_order {
                if let Some(step) = remaining_steps.remove(&step_id) {
                    execution_order.push(step);
                    executed.insert(step_id);
                    progress_made = true;
                }
            }

            if !progress_made {
                return Err("Unable to resolve step dependencies".to_string());
            }
        }

        Ok(execution_order)
    }
}

impl WorkflowStep {
    /// Create a new workflow step
    pub fn new(name: String, operation: McpOperation) -> Self {
        Self {
            id: Uuid::new_v4(),
            name,
            description: None,
            enabled: true,
            continue_on_error: false,
            timeout_ms: None,
            depends_on: Vec::new(),
            operation,
            variable_extracts: Vec::new(),
            assertions: Vec::new(),
            execution_order: 0,
        }
    }
}

File: src/lib/types/collections.ts
/**
 * World-Class Collections System for MCP Studio
 *
 * This type system enables:
 * - Cross-server workflows with variable passing
 * - Operation chaining and dependencies
 * - Test scenario persistence and replay
 * - Advanced validation and assertions
 */

import type { ServerConfig } from '../stores/serverStore';

// =============================================================================
// Core Collection Types
// =============================================================================

export interface Collection {
  id: string;
  name: string;
  description?: string;
  tags: string[];

  // Workflow definition
  workflow: WorkflowStep[];

  // Global variables and configuration
  variables: Record<string, CollectionVariable>;
  environment: CollectionEnvironment;

  // Metadata
  created_at: string;
  updated_at: string;
  created_by?: string;
  version: string;

  // Execution history
  last_run?: string;
  run_count: number;
}

export interface CollectionVariable {
  name: string;
  description?: string;
  type: 'string' | 'number' | 'boolean' | 'json' | 'secret';
  default_value?: any;
  required: boolean;

  // For runtime values
  current_value?: any;
}

export interface CollectionEnvironment {
  name: string;
  description?: string;
  servers: Record<string, string>; // alias -> server_id mapping
  variables: Record<string, any>;
}

// =============================================================================
// Workflow System
// =============================================================================

export interface WorkflowStep {
  id: string;
  name: string;
  description?: string;

  // Step configuration
  enabled: boolean;
  continue_on_error: boolean;
  timeout_ms?: number;

  // Dependencies
  depends_on: string[]; // Other step IDs that must complete first

  // The actual operation
  operation: McpOperation;

  // Variable management
  variable_extracts: VariableExtract[];
  assertions: Assertion[];

  // Execution metadata
  execution_order: number;
}

export type McpOperation =
  | ToolOperation
  | ResourceOperation
  | PromptOperation
  | SamplingOperation
  | ElicitationOperation
  | DelayOperation
  | ConditionalOperation;

// =============================================================================
// MCP Operation Types
// =============================================================================

export interface ToolOperation {
  type: 'tool';
  server_alias: string; // References Collection.environment.servers
  tool_name: string;
  parameters: Record<string, any>; // Supports variable interpolation: "${variable_name}"
}

export interface ResourceOperation {
  type: 'resource';
  server_alias: string;
  resource_uri: string; // Supports variable interpolation
}

export interface PromptOperation {
  type: 'prompt';
  server_alias: string;
  prompt_name: string;
  parameters: Record<string, string>; // MCP spec requires string values only
}

export interface SamplingOperation {
  type: 'sampling';
  server_alias: string;
  messages: Array<{
    role: 'user' | 'assistant' | 'system';
    content: string; // Supports variable interpolation
  }>;
  max_tokens?: number;
  temperature?: number;
  auto_approve?: boolean; // For automated testing
}

export interface ElicitationOperation {
  type: 'elicitation';
  server_alias: string;
  message: string; // Message to show user
  requested_schema: ElicitationSchema; // Form structure (MCP restricted schema)
  auto_approve?: boolean; // For automated testing
  default_values?: Record<string, string | number | boolean>; // Pre-fill form values
}

// MCP-compliant elicitation schema (restricted to primitives only)
export interface ElicitationSchema {
  type: 'object';
  properties: Record<string, PrimitiveSchemaDefinition>;
  required?: string[];
}

export type PrimitiveSchemaDefinition =
  | StringSchemaDefinition
  | NumberSchemaDefinition
  | BooleanSchemaDefinition
  | EnumSchemaDefinition;

export interface StringSchemaDefinition {
  type: 'string';
  title?: string;
  description?: string;
  minLength?: number;
  maxLength?: number;
  format?: 'email' | 'uri' | 'date' | 'date-time';
}

export interface NumberSchemaDefinition {
  type: 'number' | 'integer';
  title?: string;
  description?: string;
  minimum?: number;
  maximum?: number;
}

export interface BooleanSchemaDefinition {
  type: 'boolean';
  title?: string;
  description?: string;
  default?: boolean;
}

export interface EnumSchemaDefinition {
  type: 'string';
  title?: string;
  description?: string;
  enum: string[];
  enumNames?: string[]; // Display names for enum values
}

export interface DelayOperation {
  type: 'delay';
  duration_ms: number;
}

export interface ConditionalOperation {
  type: 'conditional';
  condition: string; // JavaScript expression using variables
  then_steps: string[]; // Step IDs to execute if true
  else_steps?: string[]; // Step IDs to execute if false
}

// =============================================================================
// Variable System
// =============================================================================

export interface VariableExtract {
  // Where to extract the value from
  source: 'response' | 'status' | 'timing' | 'error';

  // JSONPath or simple property access
  path: string; // e.g., "$.result.token" or "response.user.id"

  // Variable to store the extracted value
  variable_name: string;

  // Optional transformations
  transform?: VariableTransform;
}

export interface VariableTransform {
  type: 'string' | 'number' | 'boolean' | 'json' | 'base64_encode' | 'base64_decode' | 'hash';
  options?: Record<string, any>;
}

// =============================================================================
// Assertion System
// =============================================================================

export interface Assertion {
  id: string;
  name: string;
  description?: string;

  // What to assert
  type: AssertionType;

  // The assertion logic
  condition: AssertionCondition;

  // Behavior on failure
  severity: 'error' | 'warning' | 'info';
  continue_on_failure: boolean;
}

export type AssertionType =
  | 'response_status'
  | 'response_contains'
  | 'response_equals'
  | 'response_json_path'
  | 'timing'
  | 'variable_value'
  | 'custom_script';

export interface AssertionCondition {
  operator: 'equals' | 'not_equals' | 'contains' | 'not_contains' | 'greater_than' | 'less_than' | 'regex_match' | 'json_schema';
  expected_value: any;
  actual_path?: string; // JSONPath for complex response validation
}

// =============================================================================
// Execution System
// =============================================================================

export interface WorkflowExecution {
  id: string;
  collection_id: string;
  collection_name?: string;
  collection_version: string;

  // Execution metadata
  started_at: string;
  completed_at?: string;
  finished_at?: string;
  status: 'running' | 'completed' | 'failed' | 'cancelled' | 'paused';

  // Results
  step_results: Record<string, StepResult>;
  final_variables: Record<string, any>;
  summary: ExecutionSummary;

  // Runtime configuration
  environment_name: string;
  user_variables: Record<string, any>;

  // Real-time events
  events?: WorkflowExecutionEvent[];
}

export interface WorkflowExecutionEvent {
  execution_id: string;
  event_type: 'step_started' | 'step_completed' | 'step_failed' | 'execution_completed' | 'execution_failed' | 'execution_paused' | 'variable_updated';
  timestamp: string;
  step_id?: string;
  message?: string;
  data?: any;
}

export interface StepResult {
  step_id: string;
  step_name?: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'skipped';

  // Operation details
  operation_type?: string;
  operation_target?: string;

  // Timing
  started_at?: string;
  completed_at?: string;
  finished_at?: string;
  duration_ms?: number;

  // Results
  operation_result?: any;
  result?: any;
  extracted_variables: Record<string, any>;
  assertion_results: AssertionResult[];

  // Error handling
  error?: string;
  retry_count: number;
}

export interface AssertionResult {
  assertion_id: string;
  passed: boolean;
  message: string;
  expected?: any;
  actual?: any;
}

export interface ExecutionSummary {
  total_steps: number;
  completed_steps: number;
  failed_steps: number;
  skipped_steps: number;

  total_duration_ms: number;
  total_assertions: number;
  passed_assertions: number;
  failed_assertions: number;

  servers_used: string[];
  operations_performed: Record<string, number>; // operation_type -> count
}

// =============================================================================
// Import/Export System
// =============================================================================

export interface CollectionExport {
  format_version: string;
  exported_at: string;
  exported_by?: string;

  collection: Collection;
  environments: CollectionEnvironment[];

  // Optional: Include recent execution history
  recent_executions?: WorkflowExecution[];
}

export interface CollectionImportOptions {
  merge_environments: boolean;
  override_existing: boolean;
  preserve_ids: boolean;

  // Server mapping for when server IDs don't match
  server_mapping?: Record<string, string>;
}

// =============================================================================
// Collection Templates System
// =============================================================================

export interface CollectionTemplate {
  id: string;
  name: string;
  description: string;
  category: string;
  tags: string[];

  // Template-specific fields
  author?: string;
  documentation_url?: string;
  required_server_types: string[]; // e.g., ["auth-server", "api-server"]

  // The template collection (with placeholder values)
  template: Omit<Collection, 'id' | 'created_at' | 'updated_at'>;

  // Instructions for users
  setup_instructions: string;
  usage_examples: string[];
}

// =============================================================================
// UI State Management
// =============================================================================

export interface CollectionUIState {
  // Current collection being edited/viewed
  active_collection_id?: string;

  // Editor state
  selected_step_id?: string;
  editor_mode: 'visual' | 'json';

  // Execution state
  current_execution_id?: string;
  execution_paused: boolean;

  // UI preferences
  show_variable_panel: boolean;
  show_assertion_panel: boolean;
  auto_save_enabled: boolean;
}

// =============================================================================
// Type Guards and Utilities
// =============================================================================

export function isToolOperation(op: McpOperation): op is ToolOperation {
  return op.type === 'tool';
}

export function isResourceOperation(op: McpOperation): op is ResourceOperation {
  return op.type === 'resource';
}

export function isPromptOperation(op: McpOperation): op is PromptOperation {
  return op.type === 'prompt';
}

export function isSamplingOperation(op: McpOperation): op is SamplingOperation {
  return op.type === 'sampling';
}

export function isElicitationOperation(op: McpOperation): op is ElicitationOperation {
  return op.type === 'elicitation';
}

// =============================================================================
// Constants
// =============================================================================

export const COLLECTION_FORMAT_VERSION = '1.0.0';

export const DEFAULT_COLLECTION_ENVIRONMENT: CollectionEnvironment = {
  name: 'Default',
  description: 'Default environment',
  servers: {},
  variables: {}
};

export const SUPPORTED_VARIABLE_TYPES = ['string', 'number', 'boolean', 'json', 'secret'] as const;

export const SUPPORTED_ASSERTION_OPERATORS = [
  'equals', 'not_equals', 'contains', 'not_contains',
  'greater_than', 'less_than', 'regex_match', 'json_schema'
] as const;